### PR TITLE
Improve `write.GSD` and `get_snapshot` performance.

### PR DESCRIPTION
**Pull request by joaander**
## Description

* Rewrite GSD writer to directly use local particle properties in place of `takeSnapshot`.
* Rewrite serial `takeSnapshot` code path to improve performance.
* The MPI code path in `takeSnapshot` remains slow for now, but could later be rewritten with code similar to the new code in `GSDDumpWriter`.
* Topology (bond, angle, ...) snapshots also remain slow (most use cases only write this in the first frame).
* Remove `ExternalCallback` code, as it is unused and relies on the now removed `std::map` in the snapshot.
* Expand the `dynamic` flag to take in a list of individual particle fields. Keep the existing `'property'`, `'attribute'`, and `'momentum'` meta-fields for backward compatibility.

## Motivation and context

`write.GSD` is extremely slow in current releases. On profiling, I found the majority of the time was spent in creating and iterating through the `std::map` used by `takeSnapshot`. Next after that is std::vector allocations in the snapshot objects.

* The new GSD writer accesses only the particle data needed given the user parameters (e.g. in the default dynamic=['property'], only position and orientation are accessed). 
* GSD now only sorts the non-default values it will write. 
* GSD now holds on to the `std::vector` instances from one frame to the next to avoid reallocation costs.
* In MPI, GSD now uses an optimized k-way merge sort to order the particles in increasing tag order.

The optimizations to `takeSnapshot` are in this pull request because I took those steps first in an effort to boost GSD performance. There is more work to make `takeSnapshot` as fast in MPI as GSD is, but I leave that work for a future pull request.

In log only GSD files (`filter=hoomd.filter.Null()`), performance is now ~1000x faster (speedup depends on system size) as GSD no longer accesses any particle data in this case - before it always called `takeSnapshot`.

## How has this been tested?

Existing unit tests pass.

Will post benchmark results when ready to review.

## Change log

```
Changed:

* Increase performance of ``hoomd.write.GSD``.
* Increase performance of ``hoomd.State.get_snapshot`` in serial.
* `hoomd.write.GSD.dynamic` now allows fine grained control over individual particle fields.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Comment by joaander** [comment]
Benchmark results on `cheme-hodges` (NVME SSD):
* 1 rank, N=64000
  * This branch: 241 MiB/s
  * `trunk-major`: 34.8 MiB/s
* 1 rank, N=1000000
  * This branch: 203 MiB/s
  * `trunk-major`: 33.2 MiB/s
* 6 ranks, N=64000:
  * This branch: 150 MiB/s
  * `trunk-major`: 13.1 MiB/s
* 6 ranks, N=1000000)
  * This branch: 191 MiB/s
  * `trunk-major`: 11.1 MiB/s

I will add this benchmark to `hoomd-benchmarks` soon. To test, I ran` python3 -m hoomd_benchmarks.write_gsd --device CPU -v --repeat=5 --warmup_steps=100 --benchmark_steps=500 -N <N> --bandwidth`.

**Comment by joaander** [comment]
Benchmark results on `summit` (Alpine):
* 1 rank, N=64000
  * This branch: 12.2 MiB/s (16.5 frames/second)
  * `trunk-major`: 8.7 MiB/s (11.6 frames/second)
* 1 rank, N=1000000
  * This branch:  84.4 MiB/s (7.5 frames/second)
  * `trunk-major`: 14.0 MiB/s (1.28 frames/second)
* 6 ranks, N=64000:
  * This branch: 11.5 MiB/s
  * `trunk-major`:  4.79 MiB/s
* 6 ranks, N=1000000)
  * This branch: 81.6 MiB/s
  * `trunk-major`:  6.02 MiB/s

**Comment by joaander** [comment]
With the new fine grained dynamic control, setting `dynamic=['particles/position']` offers a slight performance increase over the default `dynamic=['property']` (which includes orientation).

**Comment by joaander** [comment]
@b-butler @tommy-waltmann I ask you both to review as this is a critical piece of code. Exiting unit test pass. Adding new unit tests that cover all 1024+ combinations of parameters is not feasible.

File: hoomd/GSDDumpWriter.cc [hunk 1]
```diff
@@ -150,85 +299,61 @@ void GSDDumpWriter::analyze(uint64_t timestep)
...
+        writeFrameHeader(m_local_frame);
+        writeAttributes(m_local_frame);
+        writeProperties(m_local_frame);
+        writeMomenta(m_local_frame);
         }
 
     // topology is only meaningful if this is the all group
     if (m_group->getNumMembersGlobal() == m_pdata->getNGlobal()
-        && (m_write_topology || nframes == 0))
+        && (m_write_topology || m_nframes == 0))
```
Comment (b-butler) [inline hoomd/GSDDumpWriter.cc:346]: Sort of tangential, but if we are not going to follow a user's specification (even if it doesn't make sense) , shouldn't we error rather than silently not follow his/her specification?

File: hoomd/GSDDumpWriter.cc [hunk 2]
```diff
@@ -1110,6 +1015,482 @@ void GSDDumpWriter::populateNonDefault()
...
+                {
+                image = h_image.data[index];
+                }
+
+            frame.global_box.wrap(position, image);
+
+            if (m_particle_dynamic[gsd_flag::position] || m_nframes == 0)
+                {
+                if (position.x != Scalar(0.0) || position.y != Scalar(0.0)
+                    || position.z != Scalar(0.0))
```
Comment (b-butler) [inline hoomd/GSDDumpWriter.cc]:
```suggestion
                if (position != vec3<Scalar>{0.0, 0.0, 0.0})
```
Is there a reason not to use the overloaded `==`?

File: hoomd/GSDDumpWriter.cc [hunk 3]
```diff
@@ -1110,6 +1015,482 @@ void GSDDumpWriter::populateNonDefault()
...
+                    {
+                    all_default[gsd_flag::position] = false;
+                    }
+
+                frame.particle_data.pos.push_back(vec3<float>(position));
+                }
+
+            if (m_particle_dynamic[gsd_flag::image] || m_nframes == 0)
+                {
+                if (image.x != 0 || image.y != 0 || image.z != 0)
```
Comment (b-butler) [inline hoomd/GSDDumpWriter.cc]:
```suggestion
                if (image.x != int3{0.0, 0.0, 0.0})
```

File: hoomd/GSDDumpWriter.cc [hunk 4]
```diff
@@ -1110,6 +1015,482 @@ void GSDDumpWriter::populateNonDefault()
...
+                    {
+                    all_default[gsd_flag::mass] = false;
+                    }
+
+                frame.particle_data.mass.push_back(mass);
+                }
+
+            if (m_particle_dynamic[gsd_flag::velocity] || m_nframes == 0)
+                {
+                if (velocity.x != 0.0f || velocity.y != 0.0f || velocity.z != 0.0f)
```
Comment (b-butler) [inline hoomd/GSDDumpWriter.cc]:
```suggestion
                if (velocity.x != vec3<float>{0.0, 0.0, 0.0f})
```

File: hoomd/GSDDumpWriter.cc [hunk 5]
```diff
@@ -1110,6 +1015,482 @@ void GSDDumpWriter::populateNonDefault()
...
+                                       access_location::host,
+                                       access_mode::read);
+
+        frame.particle_data_present[gsd_flag::inertia] = true;
+
+        for (unsigned int index : m_index)
+            {
+            vec3<float> inertia = vec3<float>(h_inertia.data[index]);
+
+            if (inertia.x != 0.0f || inertia.y != 0.0f || inertia.z != 0.0f)
```
Comment (b-butler) [inline hoomd/GSDDumpWriter.cc]:
```suggestion
            if (inertia != vec3<float>{0.0, 0.0, 0.0})
```

File: hoomd/GSDDumpWriter.cc [hunk 6]
```diff
@@ -1110,6 +1015,482 @@ void GSDDumpWriter::populateNonDefault()
     gsd_close(&m_handle);
     }
 
+void GSDDumpWriter::populateLocalFrame(GSDDumpWriter::GSDFrame& frame, uint64_t timestep)
```
Comment (b-butler) [inline hoomd/GSDDumpWriter.cc:1018]: Why the decision to iterate over index N attribute times rather than once with conditionals inside? Is it faster or just cleaner?

File: hoomd/md/pytest/test_gsd.py [hunk 7]
```diff
@@ -237,7 +237,7 @@ def test_write_gsd_dynamic(simulation_factory, create_md_sim, tmp_path):
 
     sim = create_md_sim
 
-    # test default dynamic=['property']
+    # test default dynamic=['property', 'property']
```
Comment (b-butler) [inline hoomd/md/pytest/test_gsd.py]:
```suggestion
    # test default dynamic=['property']
```

File: hoomd/md/pytest/test_gsd.py [hunk 8]
```diff
@@ -107,7 +107,7 @@ def test_write_gsd_trigger(create_md_sim, tmp_path):
     gsd_writer = hoomd.write.GSD(filename=filename,
```
Comment (b-butler) [inline hoomd/md/pytest/test_gsd.py:107]: While I agree that we should/can not test all combinations. It would be worth testing once using the individual attributes rather than the group ones.

File: hoomd/write/gsd.py [hunk 9]
```diff
@@ -80,37 +80,37 @@ class GSD(Writer):
...
 
     * **momentum**
 
-        * particles/velocity
-        * particles/angmom (*only written when values are not the
-          default: [0,0,0,0]*)
-        * particles/image (*only written when values are not the
-          default: [0,0,0]*)
+      * particles/velocity
+      * particles/angmom (*only written when values are not the
```
Comment (b-butler) [inline hoomd/write/gsd.py]: If we are going to document when we don't write them should we not do this for all of them?

File: hoomd/write/gsd.py
Comment (tommy-waltmann) [inline hoomd/write/gsd.py:1]: We should add something to the docstring saying that individual particle quantities can be made dynamic a la carte. Also, line 76 seems to cut off in the middle of a sentence.

File: hoomd/write/gsd.py [hunk 10]
```diff
@@ -80,37 +80,37 @@ class GSD(Writer):
```
Comment (tommy-waltmann) [inline hoomd/write/gsd.py:81]: The line above this one says that *property* is always dynamic, but I don't think that is true anymore.

File: hoomd/md/pytest/test_gsd.py
Comment (tommy-waltmann) [inline hoomd/md/pytest/test_gsd.py:1]: Let's add at least a few test cases where individual quantities are made dynamic. For example, just making `particles/velocity` dynamic instead of all of the momentum quantities, or just `particles/position` instead of all of the property quantities. I imagine these will be relatively frequent use cases.

File: hoomd/hpmc/ExternalCallback.h
Comment (tommy-waltmann) [inline hoomd/hpmc/ExternalCallback.h:1]: What does this have to do with improving `gsd` and `get_snapshot` performance?

**Review by tommy-waltmann — CHANGES_REQUESTED** [review]
This is the start of my review, I'll go through the guts of the changes in the future.

**Review by b-butler — CHANGES_REQUESTED** [review]
We need to document the setting of individual quantities in gsd.py.

Besides that and my other comments, the code looks good and well written.

File: hoomd/write/gsd.py
Comment (joaander) [reply hoomd/write/gsd.py:1]: I updated the docstring `gsd.py`.

File: hoomd/write/gsd.py (hunk 10 above)
Comment (joaander) [reply hoomd/write/gsd.py:81]: Fixed in the docstring update.

File: hoomd/hpmc/ExternalCallback.h
Comment (joaander) [reply hoomd/hpmc/ExternalCallback.h:1]: As stated in the pull request description:
> Remove `ExternalCallback` code, as it is unused and relies on the now removed `std::map` in the snapshot.

My options were to either refactor the unused and untested code, or remove it.

File: hoomd/GSDDumpWriter.cc (hunk 1 above)
Comment (joaander) [reply hoomd/GSDDumpWriter.cc:346]: It should not be an error to write the first frame of the file with a filter that selects a subset. Otherwise, users could never select a subset.

Do you think it would be useful for users to receive an exception on the 2nd frame written to the file if they both select a subset and include `'topology'` in the dynamic list?

File: hoomd/GSDDumpWriter.cc (hunk 2 above)
Comment (joaander) [reply hoomd/GSDDumpWriter.cc]: I forget that these exist.

File: hoomd/hpmc/ExternalCallback.h
Comment (tommy-waltmann) [reply hoomd/hpmc/ExternalCallback.h:1]: Apologies for that, I should have read closer.

File: hoomd/GSDDumpWriter.cc (hunk 6 above)
Comment (joaander) [reply hoomd/GSDDumpWriter.cc:1018]: Instantiating all of the `ArrayHandle`s outside the scope of one loop would require that all particle data is copied from the GPU to the CPU. Instantiating the `ArrayHandle`s inside the conditional scope in an inner loop would add excessive overhead with all the calls to `GPUArray::acquire`. Multiple loops over `m_index` is also not ideal, but I thought it to be the best of the available options.

File: hoomd/write/gsd.py (hunk 9 above)
Comment (joaander) [reply hoomd/write/gsd.py]: I pushed an update to `gsd.py` just after this review that removed all these notes. Let me know if you prefer that we instead document all the defaults here instead.

File: hoomd/md/pytest/test_gsd.py (hunk 8 above)
Comment (joaander) [reply hoomd/md/pytest/test_gsd.py:107]: Implemented.

File: hoomd/md/pytest/test_gsd.py
Comment (joaander) [reply hoomd/md/pytest/test_gsd.py:1]: Done.

File: hoomd/GSDDumpWriter.cc (hunk 1 above)
Comment (b-butler) [reply hoomd/GSDDumpWriter.cc:346]: @joaander I was thinking of cases where users specify to write out `topology` that is explicitly in `dynamic`, but the filter excludes some particles. So only erroring when `m_write_topology == true` and `m_group->getNumMembersGlobal() != m_pdata->getNGlobal()`

The error doesn't need to be here, but could be put in `setDynamic`.

File: hoomd/write/gsd.py (hunk 9 above)
Comment (b-butler) [reply hoomd/write/gsd.py]: I am good with the removal.

File: hoomd/write/gsd.py [hunk 11]
```diff
@@ -70,47 +70,44 @@ class GSD(Writer):
...
+    writes non-dynamic fields only the first frame. When reading a GSD file,
     the data in frame 0 is read when a quantity is missing in frame *i*,
-    supplying data that is static over the entire trajectory.  Set the *dynamic*
+    supplying data that is static over the entire trajectory.
 
-    Specify the one or more of the following strings in **dynamic** to make the
-    corresponding quantities dynamic (**property** is always dynamic):
+    Provide one or more of the following strings in the `dynamic` list to make
+    the corresponding field dynamic (or all fields in a given category)
+    **dynamic**:
```
Comment (tommy-waltmann) [inline hoomd/write/gsd.py]: I think this is worded a bit awkwardly, the word dynamic appears twice in a row, ignoring parentheses.

File: hoomd/write/gsd.py [hunk 12]
```diff
@@ -70,47 +70,44 @@ class GSD(Writer):
     To reduce file size, `GSD` does not write properties that are set to
     defaults. When masses, orientations, angular momenta, etc... are left
     default for all particles, these fields will not take up any space in the
-    file. Additionally, `GSD` only writes *dynamic* quantities to all frames. It
-    writes non-dynamic quantities only the first frame. When reading a GSD file,
+    file. Additionally, `GSD` only writes *dynamic* fields to all frames. It
+    writes non-dynamic fields only the first frame. When reading a GSD file,
```
Comment (tommy-waltmann) [inline hoomd/write/gsd.py]: If I understand this correctly, when a user sets a given field to the default value in frame 0, it will not be written in later frames even if the user requests that it be dynamic? That seems counterintuitive to me.

**Comment by joaander** [comment]
And on Frontier (Orion):
* 1 rank, N=64000
  * This branch: 11.64 MiB/s (15.5 frames/s)
* 1 rank, N=1000000
  * This branch:  99.8 MiB/s
* 6 ranks, N=64000:
  * This branch: 10.75 MiB/s (8.5 frames/s)
* 6 ranks, N=1000000)
  * This branch: 101 MiB/s

I have confirmed that the per-frame `fsync` for stability limits the performance on these massive scale systems to ~10-15 frames/s. When I remove the `fsync`, the benchmark achieves ~500-600 MiB/s regardless of system size.

File: hoomd/GSDDumpWriter.cc (hunk 1 above)
Comment (joaander) [reply hoomd/GSDDumpWriter.cc:346]: I don't think the error can be checked in `setDynamic`. The user may provide group that changes membership from one frame to the next.

File: hoomd/GSDDumpWriter.h [hunk 13]
```diff
@@ -153,49 +124,109 @@ class PYBIND11_EXPORT GSDDumpWriter : public Analyzer
...
-    bool m_write_momentum = false;  //!< True if momenta should be written
-    bool m_write_topology = false;  //!< True if topology should be written
-    bool m_write_diameter = false;  //!< True if the diameter attribute should be written
-    gsd_handle m_handle;            //!< Handle to the file
+    std::string m_fname;           //!< The file name we are writing to
+    std::string m_mode;            //!< The file open mode
+    bool m_truncate = false;       //!< True if we should truncate the file on every analyze()
+    bool m_is_initialized = false; //!< True if the file is open
+    bool m_write_topology = false; //!< True if topology should be written
+    bool m_write_diameter = false; //!< True if the diameter attribute should be written
```
Comment (tommy-waltmann) [inline hoomd/GSDDumpWriter.h:201]: Why do we still carry around a flag specifically for particle diameter instead of checking `m_particle_dynamic[gsd_flag::diameter]`?

File: hoomd/GSDDumpWriter.cc [hunk 14]
```diff
@@ -63,63 +64,211 @@ GSDDumpWriter::GSDDumpWriter(std::shared_ptr<SystemDefinition> sysdef,
...
+#ifdef ENABLE_MPI
+    if (m_sysdef->isDomainDecomposed())
+        {
+        m_gather_tag_order = GatherTagOrder(m_exec_conf->getMPICommunicator());
+        }
+#endif
+
+    m_particle_dynamic.reset();
+    m_particle_dynamic[gsd_flag::position] = true;
+    m_particle_dynamic[gsd_flag::orientation] = true;
```
Comment (tommy-waltmann) [inline hoomd/GSDDumpWriter.cc:77]: According to the migration guide, these are not dynamic by default any more.

**Review by tommy-waltmann — CHANGES_REQUESTED** [review]
Here are a few more comments after I went through all the changes.

File: hoomd/GSDDumpWriter.cc (hunk 14 above)
Comment (joaander) [reply hoomd/GSDDumpWriter.cc:77]: I tried to be clear in the migration guide:
> `hoomd.write.GSD` no longer enforces 'property' as an always dynamic quantity

I did not change the default (see `gsd.py` line 178) or the corresponding docstring.

The migration guide covers the case where a user provides a value for `dynamic` that doesn't include `'property'`.

For example, in v3:
```
>>> write_gsd = hoomd.write.GSD(..., dynamic = ['momentum'])
>>> print(write_gsd.dynamic)
['property', 'momentum']
```

In v4:
```
>>> write_gsd = hoomd.write.GSD(..., dynamic = ['momentum'])
>>> print(write_gsd.dynamic)
['particles/velocity', 'particles/angmom', 'particles/image']
```

If you have any suggestions on how to make this more clear in the migration guide, let me know.

File: hoomd/GSDDumpWriter.h (hunk 13 above)
Comment (joaander) [reply hoomd/GSDDumpWriter.h:201]: Because I wish the user to opt-in to writing diameters on both frame 0 and subsequent frames. So far, we only have 1 user who has expressed an interest in this: #1266

File: hoomd/write/gsd.py (hunk 12 above)
Comment (joaander) [reply hoomd/write/gsd.py]: When a field is non-default in frame 0, GSD will write it in subsequent frames when the user requests it to be dynamic.

Comment (joaander) [reply hoomd/write/gsd.py]: I noted this corner case in the documentation.

File: hoomd/write/gsd.py (hunk 11 above)
Comment (joaander) [reply hoomd/write/gsd.py]: I revised this language.

**Review by b-butler — APPROVED** [review]

**Review by tommy-waltmann — APPROVED** [review]
I think the documentation is about as clear as its going to get for now. We can always go back and adjust wording based on confusion from users on the mailing list if needed.
